cai_causal_graph/identify_utils.py

```diff
@@ -0,0 +1,72 @@
+"""
+Copyright 2023 Impulse Innovations Limited
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+from typing import List
+
+from cai_causal_graph import CausalGraph
+from cai_causal_graph.exceptions import CausalGraphErrors
+
+
+def identify_confounders(graph: CausalGraph, source: str, destination: str) -> List[str]:
+    """
+    Identify all confounders between the `source` and `destination` in the provided `graph`.
+
+    A confounder between the `source` and `destination` node is a node that is a (minimal) ancestor of both the
+    `source` and the `destination` node. Being a _minimal_ ancestor here means that the node is not an ancestor of
+    other confounder nodes, unless it has another directed path to the `destination` node that does not go through other
+    confounder nodes.
+
+    Example:
+        >>> from typing import List
+        >>> from cai_causal_graph import CausalGraph
+        >>> from cai_causal_graph.identify_utils import identify_confounders
+        >>>
+        >>> # define a causal graph
+        >>> cg = CausalGraph()
+        >>> cg.add_edge('z', 'u')
+        >>> cg.add_edge('u', 'x')
+        >>> cg.add_edge('u', 'y')
+        >>> cg.add_edge('x', 'y')
+        >>>
+        >>> # compute confounders between source and destination; output: ['u']
+        >>> confounders_list: List[str] = identify_confounders(cg, source='x', destination='y')
+
+    :param graph: The causal graph given by a `cai_causal_graph.causal_graph.CausalGraph` instance. This must be a DAG,
+        i.e. it must only contain directed edges, otherwise an error is raised.
+    :param source: The source variable.
+    :param destination: The destination variable.
+    :return: A list of all confounders between the `source` and `destination`.
+    """
+    if not graph.is_dag():
+        raise TypeError(f'Expected a DAG, but got a mixed causal graph.')
+
+    # create a copy of the provided graph and prune the edge between source and destination
+    pruned_graph = graph.copy()
+    try:
+        pruned_graph.remove_edge(source=source, destination=destination)
```

Comment:
This ignores the edge type. What if we have Z -> X, Z -> Y, Y-X but say source=X, destination=Y then should we have empty list returned? However, this would return Z.

Reply:
Maybe we don't actually care source vs destination but just between 2 nodes.

Reply:
If order of 2 nodes doesn't matter then would add unit test to confirm `identify_confounders(g, 'x', 'y')` and `identify_confounders(g, 'y', 'x')` returns the same list.

Reply:
I think that's a good suggestion. I do agree that we should get the same result if the `source` and `destination` are swapped. Cheers

Reply:
Should we change the name though as there is no real idea of source vs destination? It implies one is above other in topological sort. Would maybe just say `node_1` and `node_2`.

Reply:
I think just `source` and `destination` is totally fine to be honest, but I'm happy to change it if you have a strong opinion

Reply:
I would be worried about the backward compatibility here. Maybe enough to explain in the docstring that the notion of source and destination is vague and maybe sometimes ignored in directed edges?

Reply:
If you believe it's necessary to explain this further in the docstrings, I'd rather just change the names to `node_1` and `node_2`

Reply:
Changed now!

Reply:
Thanks! @mmiklitz no concern for backwards compat as this util never existed before.